### PR TITLE
Add pipeline module and ase function

### PR DESCRIPTION
**Pull request by daxpryce**
- [ ] Does this PR add any new dependencies?
- [ ] Does this PR modify any existing APIs?
   - [ ] Is the change to the API backwards compatible?
- [x] Have you built the documentation (reference and/or tutorial) and verified the generated documentation is appropriate?

So, this is the first commits to the pipeline module.  It includes a few QOL items like an int-ifying graph builder class that will generate an `nx.Graph` or `nx.DiGraph`, a preconditions module for type checking (probably could end up in a utils module in the top level), an Embeddings return object that makes it trivial to correlate a node id to it's respective tensor, and an implementation of ASE-as-a-function.

Of particular note is that it does not use the built-in elbow finder yet, instead behaving more similarly to the topologic approach of using a combination of fixed dimensionality requested and the elbow finder on that.  We've discussed addressing our `graspologic.embed.AdjacencySpectralEmbed` elbow-finding concerns in a meeting and we're reasonably confident we can have a consistent behavior in both in the future, but for now this works well enough.

### CODE REVIEW

File: graspologic/embed/svd.py [hunk 1]
```diff
@@ -285,7 +298,12 @@ def selectSVD(X, n_components=None, n_elbows=2, algorithm="randomized", n_iter=5
         V = V[idx, :]
 
     elif algorithm == "randomized":
-        U, D, V = sklearn.utils.extmath.randomized_svd(X, n_components, n_iter=n_iter)
+        # for some reason, randomized_svd defaults random_state to 0 if not provided
+        # which is weird because None is a valid starting point too
```
Comment (bdpedigo) [inline graspologic/embed/svd.py:302]: so are you trying to stick with their convention?

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    graph: Union[nx.Graph, nx.DiGraph],
+    dimensions: int = 100,
+    elbow_cut: Optional[int] = None,
+    svd_solver_algorithm: str = "randomized",
+    svd_solver_iterations: int = 5,
+    svd_seed: Optional[int] = None,
+    weight_attribute: str = "weight",
+) -> Embeddings:
+    """
+    Given a weighted directed or undirected networkx graph (*not* multigraph),
```
Comment (bdpedigo) [inline graspologic/pipeline/embed/adjacency_spectral_embedding.py:33]: doesn't have to be weighted

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    generate an Embeddings object.
+
+    Adjacency spectral embeddings are extremely egocentric.
```
Comment (bdpedigo) [inline graspologic/pipeline/embed/adjacency_spectral_embedding.py:36]: what does egocentric mean here?

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+    `Adjacency Spectral Embedding Tutorial
+    <https://microsoft.github.io/graspologic/tutorials/embedding/AdjacencySpectralEmbed.html>`_
+
+    In addition to diagonal augmentation, all graphs will have pass to ranks executed.
```
Comment (bdpedigo) [inline graspologic/pipeline/embed/adjacency_spectral_embedding.py:41]: i feel like there is not much context for either of these statements

File: graspologic/embed/svd.py (hunk 1 above)
Comment (daxpryce) [reply graspologic/embed/svd.py:302]: I think it actually broke some unit tests when I changed it to None so I did this song and dance around it.  I'll change it to None and see what we end up with - it may be that we go with it in the short term and fix it in another PR?

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py (hunk 2 above)
Comment (bdpedigo) [reply graspologic/pipeline/embed/adjacency_spectral_embedding.py:33]: mathematically at least - you're saying it does for this function?

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+    Parameters
+    ----------
+    graph : Union[nx.Graph, nx.DiGraph]
+        A simple, weighted graph, either undirected or directed. The graph **must** be
+        fully weighted (every edge has a weight), it **must** be a simple graph (meaning
```
Comment (bdpedigo) [inline graspologic/pipeline/embed/adjacency_spectral_embedding.py:47]: simple graph may be an overloaded term here https://mathworld.wolfram.com/SimpleGraph.html#:~:text=A%20simple%20graph%2C%20also%20called,edges%20(Gibbons%201985%2C%20p.&text=A%20simple%20graph%20may%20be,refers%20to%20a%20simple%20graph.

File: graspologic/pipeline/embed/embeddings.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                f"embeddings: {embeddings.shape[0]}"
+            )
+        self._labels = labels
+        self._embeddings = embeddings
+
+    def labels(self) -> np.ndarray:
+        return self._labels
+
+    def embeddings(self) -> np.ndarray:
+        return self._embeddings
```
Comment (bdpedigo) [inline graspologic/pipeline/embed/embeddings.py:74]: would these make sense as properties?

File: graspologic/pipeline/embed/embeddings.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from typing import Any, Tuple
+import numpy as np
+from collections import OrderedDict
+
+
+class Embeddings:
```
Comment (bdpedigo) [inline graspologic/pipeline/embed/embeddings.py:23]: i will likely lose this battle but IMO a container with just an index and a numpy array of values is quite similar to a pandas dataframe, which literally has an index array and a numpy array, with the additional advantage that it is ubiquitous and easy for plotting. and if we are making someone learn our obscure data handling object vs a pandas dataframe, i'd vote for pandas.

File: graspologic/pipeline/preconditions.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (bdpedigo) [inline graspologic/pipeline/preconditions.py:1]: is there a reason to keep any of these in pipeline, as opposed to root level module?

File: graspologic/embed/svd.py (hunk 1 above)
Comment (daxpryce) [reply graspologic/embed/svd.py:302]: Yeah, it's breaking more than a few.  I think we can and should fix this, but it wasn't something I was trying to do in this PR - it already touched on too many classes, I think!

Results of commenting out line 303:

```
============================================================================= FAILURES ==============================================================================
_________________________________________________________________________ TestGMP.test_sim __________________________________________________________________________

self = <tests.test_match.TestGMP object at 0x1448ff640>

    def test_sim(self):
        n = 150
        rho = 0.9
        n_per_block = int(n / 3)
        n_blocks = 3
        block_members = np.array(n_blocks * [n_per_block])
        block_probs = np.array(
            [[0.2, 0.01, 0.01], [0.01, 0.1, 0.01], [0.01, 0.01, 0.2]]
        )
        directed = False
        loops = False
        A1, A2 = sbm_corr(
            block_members, block_probs, rho, directed=directed, loops=loops
        )
        ase = AdjacencySpectralEmbed(n_components=3, algorithm="truncated")
        x1 = ase.fit_transform(A1)
        x2 = ase.fit_transform(A2)
        xh1 = SignFlips().fit_transform(x1, x2)
        S = xh1 @ x2.T
        res = self.barygm.fit(A1, A2, S=S)

>       assert 0.7 <= (sum(res.perm_inds_ == np.arange(n)) / n)
E       assert 0.7 <= (78 / 150)
E        +  where 78 = sum(array([ 86,  ...47, 148, 149]) == array([  0,  ...47, 148, 149])
E           Use -v to get the full diff)

tests/test_match.py:251: AssertionError
___________________________________________________________________ TestDCSBM.test_DCSBM_nparams ____________________________________________________________________

self = <tests.test_models.TestDCSBM object at 0x131e2c6d0>

    def test_DCSBM_nparams(self):
        n_verts = 3000
        n_class = 4
        graph = self.g
        labels = self.labels
        e = DCSBMEstimator(directed=True)
        e.fit(graph)
>       assert e._n_parameters() == (n_verts + n_class - 1 + n_class ** 2)
E       assert 3071 == (((3000 + 4) - 1) + (4 ** 2))
E        +  where 3071 = <bound method DCSBMEstimator._n_parameters of DCSBMEstimator()>()
E        +    where <bound method DCSBMEstimator._n_parameters of DCSBMEstimator()> = DCSBMEstimator()._n_parameters

tests/test_models.py:422: AssertionError
______________________________________________________________________ TestOmni.test_diag_aug _______________________________________________________________________

self = <tests.test_omni.TestOmni testMethod=test_diag_aug>

    def test_diag_aug(self):
        np.random.seed(5)
        n = 100
        p = 0.25

        graphs_list = [er_np(n, p) for _ in range(2)]
        graphs_arr = np.array(graphs_list)

        # Test that array and list inputs results in same embeddings
        omni_arr = OmnibusEmbed(diag_aug=True).fit_transform(graphs_arr)
        omni_list = OmnibusEmbed(diag_aug=True).fit_transform(graphs_list)

>       np.testing.assert_array_equal(omni_list, omni_arr)
E       AssertionError:
E       Arrays are not equal
E
E       Mismatched elements: 1200 / 1200 (100%)
E       Max absolute difference: 1.415089
E       Max relative difference: 212.39581875
E        x: array([[[ 0.508661, -0.044478, -0.16163 ,  0.092923,  0.106579,
E                -0.216692],
E               [ 0.436264,  0.371707,  0.193453, -0.028346, -0.123348,...
E        y: array([[[ 0.508661, -0.015476, -0.238229,  0.027816,  0.019887,
E                -0.283967],
E               [ 0.436264,  0.338461,  0.145553, -0.009014, -0.131633,...

tests/test_omni.py:133: AssertionError
__________________________________________________ TestAdjacencySpectralEmbed.test_transform_closeto_fit_transform __________________________________________________

self = <tests.test_spectral_embed.TestAdjacencySpectralEmbed testMethod=test_transform_closeto_fit_transform>

    def test_transform_closeto_fit_transform(self):
        atol = 0.15
        for diag_aug in [True, False]:
            for g, A in self.testgraphs.items():
                ase = AdjacencySpectralEmbed(
                    n_components=2, diag_aug=diag_aug, svd_seed=9001
                )
                ase.fit(A)
                Y = ase.fit_transform(A)
                if isinstance(Y, np.ndarray):
                    X = ase.transform(A)
                    np.testing.assert_allclose(X, Y, atol=atol)
                elif isinstance(Y, tuple):
                    with self.assertRaises(TypeError):
                        X = ase.transform(A)
                    X = ase.transform((A.T, A))
                    np.testing.assert_allclose(X[0], Y[0], atol=atol)
>                   np.testing.assert_allclose(X[1], Y[1], atol=atol)
E                   AssertionError:
E                   Not equal to tolerance rtol=1e-07, atol=0.15
E
E                   Mismatched elements: 1 / 40 (2.5%)
E                   Max absolute difference: 0.18669627
E                   Max relative difference: 5.41864066
E                    x: array([[ 1.523434, -0.9013  ],
E                          [ 1.624558, -0.937225],
E                          [ 0.753984, -0.933442],...
E                    y: array([[ 1.586404, -0.957384],
E                          [ 1.711405, -0.99416 ],
E                          [ 0.791389, -0.990951],...

tests/test_spectral_embed.py:178: AssertionError
=======================================================================
```

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py (hunk 2 above)
Comment (daxpryce) [reply graspologic/pipeline/embed/adjacency_spectral_embedding.py:33]: I was, but we actually allow unweighted and we make them into weighted, so it's a good catch

File: graspologic/embed/svd.py (hunk 1 above)
Comment (bdpedigo) [reply graspologic/embed/svd.py:302]: so looks like many of these were just fragile to exactly what SVD they were getting out with the implicitly set randomized seed. just off the top of my head - 
 - TestMatch: it's possible this test was super fragile to the SVD seed, which we should fix?
 - DCSBM: no idea, maybe was also fragile to the SVD seed which messed up the embedding which led to wrong # of parameters?
 - Omni: can't expect exact same SVDs if using randomized and the seeds are not set, should probably just set the seed for this test.
 - ASE transform: probably should just set the seed here also for the same reason as the above.

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py (hunk 3 above)
Comment (daxpryce) [reply graspologic/pipeline/embed/adjacency_spectral_embedding.py:36]: it's a great question - I stole it from J-Lar.  He often describes ASE as "extremely egocentric" and LSE as "core periphery" but I couldn't tell you exactly what he means by that.  So we'll summon him with @JonathanLarson and see if he can shed some light on it.  I'm ok with removing it if we can't get a good justification.

Comment (bdpedigo) [reply graspologic/pipeline/embed/adjacency_spectral_embedding.py:36]: i guess my point was mainly that it seems like something we should define or have another term for in the spirit of making it developer friendly. I think is referring to something along the lines of a focus on local neighborhood of a node?

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py (hunk 5 above)
Comment (daxpryce) [reply graspologic/pipeline/embed/adjacency_spectral_embedding.py:47]: what's the short and sweet descriptor of a graph that is explicitly not a multigraph?

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py (hunk 3 above)
Comment (daxpryce) [reply graspologic/pipeline/embed/adjacency_spectral_embedding.py:36]: I agree with you, which is why I'm trying to get Jon to come in here and give us more details on why he finds that a good descriptor and exactly what it means.  I have my understanding based on colloquial meanings of ego-centric, but I don't want to say they're right.

File: graspologic/pipeline/embed/embeddings.py (hunk 6 above)
Comment (daxpryce) [reply graspologic/pipeline/embed/embeddings.py:74]: I didn't really want people to be thinking it made sense to set these at any point.  we could but I donno that it buys us anything important really?

File: graspologic/pipeline/embed/embeddings.py (hunk 7 above)
Comment (daxpryce) [reply graspologic/pipeline/embed/embeddings.py:23]: ubiquitous in the datascience domain, not in the application domain.  I'm trying to build for the latter, specifically.

File: graspologic/pipeline/preconditions.py (hunk 8 above)
Comment (daxpryce) [reply graspologic/pipeline/preconditions.py:1]: 1: I'm hoping we can get rid of the type checking after we check out whether beartype will work for us
2: yes, regardless of whether we get rid of the type checking part or not, it should go into a utils package.  Or at least be made available for the rest of graspologic.  I just wanted to set it as a trial balloon here to see if anyone would want to use it before I "graduated" it up in the package structure.  I can totally move it if you want me to now though

Comment (bdpedigo) [reply graspologic/pipeline/preconditions.py:1]: i feel like i'm for that, even if we change the structure/details later

File: graspologic/pipeline/embed/embeddings.py (hunk 6 above)
Comment (bdpedigo) [reply graspologic/pipeline/embed/embeddings.py:74]: i guess just whether you access them by obj.embeddings vs obj.embeddings() (method vs attribute-like thing) was my main point

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py (hunk 5 above)
Comment (bdpedigo) [reply graspologic/pipeline/embed/adjacency_spectral_embedding.py:47]: i guess I don't know of one other than "not multigraph"

File: graspologic/embed/svd.py (hunk 1 above)
Comment (daxpryce) [reply graspologic/embed/svd.py:302]: Something else is still weird in the ase transform - it *is* setting the seed, yet it isn't getting the same results.  we don't create a random state of our own with the seed - we use the seed as an int every time, so any time we call select_svd with randomized algorithm, we're going to use the seed value 9001.  so... this shouldn't be failing, and I'm not entirely sure what we're trying to test with it.  

This just underscores that I'm going to need someone else to look into these with more detail.  I can probably fix the omni one - I'm pretty sure I just missed that one.  The others just left me with question marks firing off like crazy.

File: graspologic/pipeline/embed/embeddings.py (hunk 6 above)
Comment (daxpryce) [reply graspologic/pipeline/embed/embeddings.py:74]: So it's basically a subtle semantics implication/problem and me trying to do my best to limit unsupported/unintended usage:
`obj.embeddings` -> as an attribute, you can access it directly, safely, and mutate it without regard for side effects (which is a terrible idea in this case)
`obj._embeddings` -> as an attribute, it's private: you are risking the Embedding integrity by mutating it, and you really shouldn't look at it
`obj._embeddings` with `obj.embeddings()` getter function: you can safely get these values to look at, but probably only through this method

Now, I know we're just passing a pointer in the latter case, and you can mutate it all you want - and you can mutate the attribute all you want too.  Python lets you do whatever you want whenever you want without guard rails.

Since the language doesn't let us limit a lot of behaviors that other languages do, we're stuck defining things that *imply* the behaviors we're wanting to support and hoping people agree to observe the rules of what is implied and don't try to ... well, do whatever crazy stuff they want to do.

File: graspologic/pipeline/embed/adjacency_spectral_embedding.py (hunk 5 above)
Comment (daxpryce) [reply graspologic/pipeline/embed/adjacency_spectral_embedding.py]: ended up going with basic graph, and if that has a term overload then I don't know.  describing something as "not something more complex" vexes me

**Review by bryantower — APPROVED** [review]

File: graspologic/embed/svd.py (hunk 1 above)
Comment (daxpryce) [reply graspologic/embed/svd.py:302]: #815 btw
